### PR TITLE
[beta]: Prepare the 1.31.0 beta release

### PR DESCRIPTION
**Pull request by alexcrichton**
* Update to Cargo's branched 1.31.0 branch
* Update the channel to `beta`

Rolled up beta-accepted PRs:

* #55362: Remove `cargo new --edition` from release notes.
* #55325: Fix link to macros chapter
* #55358: Remove redundant clone (2) 
* #55346: Shrink `Statement`.
* #55274: Handle bindings in substructure of patterns with type ascriptions
* #55297: Partial implementation of uniform paths 2.0 to land before beta
* #55192: Fix ordering of nested modules in non-mod.rs mods
* #55185: path suggestions in Rust 2018 should point out the change in semantics
* #55423: back out bogus `Ok`-wrapping suggestion on `?` arm type mismatch 

Note that **this does not update the bootstrap compiler** due to https://github.com/rust-lang/rust/issues/55404

### CODE REVIEW

**Comment by alexcrichton** [comment]
@Mark-Simulacrum you may also know best how to triage/tag https://github.com/rust-lang/rust/issues/55404

**Comment by pietroalbini** [comment]
Rolled up beta-accepted PRs inside this one.

@bors r=Mark-Simulacrum p=5

**Comment by Mark-Simulacrum** [comment]
Cc @nrc - looks like we might need an RLS update.. though this might be because of the lack of bootstrap compiler update...

@aturon @steveklabnik This probably means we can't get RC2 out tomorrow - it'll take probably till Tuesday

**Comment by Mark-Simulacrum** [comment]
Yes - I'm not sure how this managed to break. I probably won't get a chance to take a look until tomorrow night unfortunately...

**Comment by pietroalbini** [comment]
A quick fix would be to branch the beta off before those build failures happened.

**Comment by pietroalbini** [comment]
The last commit before the tools breakage is 4bd4e4130ed531a644263db26bf8461704215c77. If we start the beta with that branch, the following PRs will also need to be backported:

* #55362: Remove `cargo new --edition` from release notes. (relnotes update)
* #55325: Fix link to macros chapter (relnotes update)
* #55274: Handle bindings in substructure of patterns with type ascriptions (wanted for RC2)

I think this is the easiest way to go forward, and if we move fast we should be able to release beta this evening (possibly earlier if we manually start the job on RCS).

<details><summary>Full list of PRs between 4bd4e4130e and 10f42cbde</summary>

* ~~#54145: Keep resolved defs in path prefixes and emit them in save-analysis~~ (broke rls)
* **#55362: Remove `cargo new --edition` from release notes.** (relnotes update)
* #53821: Report const eval error inside the query
* #54626: rustc: Tweak filenames encoded into metadata
* #54816: Don't try to promote already promoted out temporaries
* #54824: Cleanup rustdoc tests with @!has and @!matches
* #54921: Add line numbers option to rustdoc
* #55167: Add a "cheap" mode for compute_missing_ctors.
* #55258: Fix Rustdoc ICE when checking blanket impls
* #55264: Compile the libstd we distribute with -Ccodegen-unit=1
* #55271: Unimplement ExactSizeIterator for MIR traversing iterators
* #55292: Macro diagnostics tweaks
* #55298: Point at macro definition when no rules expect token
* #55301: List allowed tokens after macro fragments
* #55302: Extend the impl_stable_hash_for! macro for miri.
* **#55325: Fix link to macros chapter** (relnotes update)
* #55343: rustbuild: fix remap-debuginfo when building a release
* #55346: Shrink Statement.
* #55358: Remove redundant clone: 2)
* #55370: Update mailmap for estebank
* #55375: Typo fixes in configure_cmake comments
* #55378: rustbuild: use configured linker to build boostrap
* #55379: validity: assert that unions are non-empty
* #55383: Use SmallVec for the queue in coerce_unsized.
* #55391: bootstrap: clean up a few clippy findings
* **#55274: Handle bindings in substructure of patterns with type ascriptions** (wanted for RC2)

</details>

**Comment by pietroalbini** [comment]
Also included the following PRs to avoid some conflicts while cherry-picking:

* #55358: Remove redundant clone (2)
* #55346: Shrink Statement.

Unfortunately https://github.com/rust-lang/rust/pull/55297 and https://github.com/rust-lang/rust/pull/55185 failed to cherry-pick cleanly (they depend on the broken RLS PR), I hope I didn't mess up the conflicts resolution.

@bors r+ p=1000

**Comment by pietroalbini** [comment]
WTF

File: src/librustc_resolve/lib.rs [hunk 1]
```diff
@@ -1010,7 +1010,7 @@ pub struct ModuleData<'a> {
     resolutions: RefCell<FxHashMap<(Ident, Namespace), &'a RefCell<NameResolution<'a>>>>,
     legacy_macro_resolutions: RefCell<Vec<(Ident, MacroKind, ParentScope<'a>,
                                            Option<&'a NameBinding<'a>>)>>,
-    macro_resolutions: RefCell<Vec<(Vec<Segment>, ParentScope<'a>, Span)>>,
+    macro_resolutions: RefCell<Vec<(Vec<Ident>, ParentScope<'a>, Span)>>,
```
Comment (petrochenkov) [inline src/librustc_resolve/lib.rs]: Oh, man.
Couldn't you just wait with beta for a day longer, as it was promised.

Comment (pietroalbini) [reply src/librustc_resolve/lib.rs]: @Mark-Simulacrum said they were OK with this on IRC.

Comment (pietroalbini) [reply src/librustc_resolve/lib.rs]: RC2 was announced in the blog to be released today, so we wanted to reduce the delay for beta.

Comment (petrochenkov) [reply src/librustc_resolve/lib.rs]: OK, but "resolve: Make sure macros and imports are resolved in full parent scope" was somehow cherry-picked incorrectly and "fix errors caused by cherry picking" now "fixes" something that was already correct.

**Comment by Mark-Simulacrum** [comment]
@petrochenkov FWIW I think there was a bit of miscommunication with beta promotion timeline.

**Comment by pietroalbini** [comment]
What is happening with bors today `:/`

**Comment by petrochenkov** [comment]
@Mark-Simulacrum 
Wait a second, https://github.com/rust-lang/rust/pull/54145 is not in beta, how is that even possible?
Beta was branched off Oct 25, 2018, 4 days ago?!
That's even before this comment https://github.com/rust-lang/rust/pull/55297#issuecomment-432482537.

That's not a bit of miscommunication, that's a total failure of communication.
All that stuff was explicitly targeting beta and worked on with that assumption.

**Comment by petrochenkov** [comment]
I see now what's the source of the cherry-picking issues - https://github.com/rust-lang/rust/pull/54145 was supposed to be on beta, and https://github.com/rust-lang/rust/pull/55297 and https://github.com/rust-lang/rust/pull/55185 were supposed to go after it, but as it turns out https://github.com/rust-lang/rust/pull/54145 is not in fact on beta.

**Comment by petrochenkov** [comment]
@pietroalbini 
Oh. I see.
If https://github.com/rust-lang/rust/pull/54145 is indeed "un-betable", then this PR probably did the best it could.
I guess I'll have to continue work on https://github.com/rust-lang/rust/pull/55297 directly on top of beta in this case.

**Comment by pietroalbini** [comment]
@petrochenkov I think we can backport #54145 once RLS is fixed (along with the RLS fix), so I wouldn't worry about working on top of beta. @Mark-Simulacrum can confirm that though.

**Pull request merged**